filters/auth/grant.go

```diff
@@ -0,0 +1,197 @@
+package auth
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+	"golang.org/x/oauth2"
+)
+
+const (
+	OAuthGrantName = "oauthGrant"
+
+	secretsRefreshInternal    = time.Minute
+	oauthGrantRefreshTokenKey = "oauth-grant-token"
+)
+
+var (
+	errExpiredToken = errors.New("expired access token")
+)
+
+type grantSpec struct {
+	config OAuthConfig
+}
+
+type grantFilter struct {
+	config OAuthConfig
+}
+
+func (s *grantSpec) Name() string { return OAuthGrantName }
+
+func (s *grantSpec) CreateFilter([]interface{}) (filters.Filter, error) {
+	return &grantFilter{
+		config: s.config,
+	}, nil
+}
+
+func providerContext(c OAuthConfig) context.Context {
+	return context.WithValue(context.Background(), oauth2.HTTPClient, c.AuthClient)
+}
+
+func serverError(ctx filters.FilterContext) {
+	ctx.Serve(&http.Response{
+		StatusCode: http.StatusInternalServerError,
+	})
+}
+
+func badRequest(ctx filters.FilterContext) {
+	ctx.Serve(&http.Response{
+		StatusCode: http.StatusBadRequest,
+	})
+}
+
+func loginRedirect(ctx filters.FilterContext, config OAuthConfig) {
+	loginRedirectWithOverride(ctx, config, "")
+}
+
+func loginRedirectWithOverride(ctx filters.FilterContext, config OAuthConfig, originalOverride string) {
+	req := ctx.Request()
+	redirect, original := config.RedirectURLs(req)
+
+	if originalOverride != "" {
+		original = originalOverride
+	}
+
+	state, err := config.flowState.createState(original)
+	if err != nil {
+		log.Errorf("Failed to create login redirect: %v", err)
+		serverError(ctx)
+		return
+	}
+
+	authConfig := config.GetConfig()
+	ctx.Serve(&http.Response{
+		StatusCode: http.StatusTemporaryRedirect,
+		Header: http.Header{
+			"Location": []string{authConfig.AuthCodeURL(state, config.GetAuthURLParameters(redirect)...)},
+		},
+	})
+}
+
+func (f *grantFilter) refreshToken(c cookie) (*oauth2.Token, error) {
+	// Set the expiry of the token to the past to trigger oauth2.TokenSource
+	// to refresh the access token.
+	token := &oauth2.Token{
+		AccessToken:  c.AccessToken,
+		RefreshToken: c.RefreshToken,
+		Expiry:       time.Now().Add(-time.Minute),
+	}
+
+	ctx := providerContext(f.config)
+
+	// oauth2.TokenSource implements the refresh functionality,
+	// we're hijacking it here.
+	tokenSource := f.config.GetConfig().TokenSource(ctx, token)
+	return tokenSource.Token()
+}
+
+func (f *grantFilter) refreshTokenIfRequired(c cookie) (*oauth2.Token, error) {
+	now := time.Now()
+	isAccessTokenExpired := c.isAccessTokenExpired()
+	canRefresh := !f.config.DisableRefresh && c.RefreshToken != ""
+	shouldRefresh := isAccessTokenExpired || now.After(c.RefreshAfter)
```

Comment:
not sure if we need the `RefreshAfter` clause, because we can just use the RefreshToken in case AccessToken is expired, see also https://github.com/zalando/skipper/pull/1609#discussion_r531645346
Do I miss something?

Reply:
after the linked considerations, this only depends on whether we see any value in refreshing the access token early on.

Reply:
Ok then let's drop it!